File: packages/steamvr-lighthouse-driver-gamepads/src/pose.js
const { multiplyQuaternions, vector3Add, applyQuaternion } = require('./math')
const { new_VRPose_withEverything, new_VRPose_from } = require('webvr-types-constructors')

// openvr_driver.h
// /* Generally, the pose maintained by a driver
//  * is in an inertial coordinate system different
//  * from the world system of x+ right, y+ up, z+ back.
//  * Also, the driver is not usually tracking the "head" position,
//  * but instead an internal IMU or another reference point in the HMD.
//  * The following two transforms transform positions and orientations
//  * to app world space from driver world space,
//  * and to HMD head space from driver local body space. 
//  *
//  * We maintain the driver pose state in its internal coordinate system,
//  * so we can do the pose prediction math without having to
//  * use angular acceleration.  A driver's angular acceleration is generally not measured,
//  * and is instead calculated from successive samples of angular velocity.
//  * This leads to a noisy angular acceleration values, which are also
//  * lagged due to the filtering required to reduce noise to an acceptable level.
//  */

//ISSUE we're throwing away DriverPose_t's poseTimeOffset.
//Perhaps timeStamp = performance.now - poseTimeOffset?
// But that would mean trusting it.


function worldPoseFromDevicePose (pose,dp) {
  vector3Add(pose.position, dp.vecPosition, dp.vecWorldFromDriverTranslation);
  multiplyQuaternions(pose.orientation, dp.qRotation, dp.qWorldFromDriverRotation);
  applyQuaternion(pose.linearVelocity, dp.vecVelocity, dp.qWorldFromDriverRotation);
  applyQuaternion(pose.linearAcceleration, dp.vecAcceleration, dp.qWorldFromDriverRotation);
  applyQuaternion(pose.angularVelocity, dp.vecAngularVelocity, dp.qWorldFromDriverRotation);
  applyQuaternion(pose.angularAcceleration, dp.vecAngularAcceleration, dp.qWorldFromDriverRotation);
}

function setPose (gamepad,device) {
  var pose = null;
  if (device.pose) {
    gamepad.connected = device.pose.deviceIsConnected;
    if (device.pose.poseIsValid) {
      pose = new_VRPose_withEverything();
      worldPoseFromDevicePose(pose,device.pose);
    }
  }
  //ISSUE Gamepad.pose isn't spec as of 2016-09-30.
  gamepad.pose = pose;
}

module.exports = {
  setPose
}

File: packages/steamvr-lighthouse-driver/src/addon/pose.cc
#include "internal.h"

v8::Local<v8::Object> objectFromPose (vr::DriverPose_t pose) {
  v8::Local<v8::Array> arr;
  v8::Local<v8::Object> obj = Nan::New<v8::Object>();

  obj->Set(Nan::New("timeOffset").ToLocalChecked(), Nan::New(pose.poseTimeOffset));
  
  arr = Nan::New<v8::Array>(4);
  arr->Set(0, Nan::New(pose.qWorldFromDriverRotation.x));
  arr->Set(1, Nan::New(pose.qWorldFromDriverRotation.y));
  arr->Set(2, Nan::New(pose.qWorldFromDriverRotation.z));
  arr->Set(3, Nan::New(pose.qWorldFromDriverRotation.w));
  obj->Set(Nan::New("qWorldFromDriverRotation").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecWorldFromDriverTranslation[0]));
  arr->Set(1, Nan::New(pose.vecWorldFromDriverTranslation[1]));
  arr->Set(2, Nan::New(pose.vecWorldFromDriverTranslation[2]));
  obj->Set(Nan::New("vecWorldFromDriverTranslation").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(4);
  arr->Set(0, Nan::New(pose.qDriverFromHeadRotation.x));
  arr->Set(1, Nan::New(pose.qDriverFromHeadRotation.y));
  arr->Set(2, Nan::New(pose.qDriverFromHeadRotation.z));
  arr->Set(3, Nan::New(pose.qDriverFromHeadRotation.w));
  obj->Set(Nan::New("qDriverFromHeadRotation").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecDriverFromHeadTranslation[0]));
  arr->Set(1, Nan::New(pose.vecDriverFromHeadTranslation[1]));
  arr->Set(2, Nan::New(pose.vecDriverFromHeadTranslation[2]));
  obj->Set(Nan::New("vecDriverFromHeadTranslation").ToLocalChecked(), arr);

  // right, up, back
  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecPosition[0]));
  arr->Set(1, Nan::New(pose.vecPosition[1]));
  arr->Set(2, Nan::New(pose.vecPosition[2]));
  obj->Set(Nan::New("vecPosition").ToLocalChecked(), arr);
  
  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecVelocity[0]));
  arr->Set(1, Nan::New(pose.vecVelocity[1]));
  arr->Set(2, Nan::New(pose.vecVelocity[2]));
  obj->Set(Nan::New("vecVelocity").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecAcceleration[0]));
  arr->Set(1, Nan::New(pose.vecAcceleration[1]));
  arr->Set(2, Nan::New(pose.vecAcceleration[2]));
  obj->Set(Nan::New("vecAcceleration").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(4);
  arr->Set(0, Nan::New(pose.qRotation.x));
  arr->Set(1, Nan::New(pose.qRotation.y));
  arr->Set(2, Nan::New(pose.qRotation.z));
  arr->Set(3, Nan::New(pose.qRotation.w));
  obj->Set(Nan::New("qRotation").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecAngularVelocity[0]));
  arr->Set(1, Nan::New(pose.vecAngularVelocity[1]));
  arr->Set(2, Nan::New(pose.vecAngularVelocity[2]));
  obj->Set(Nan::New("vecAngularVelocity").ToLocalChecked(), arr);

  arr = Nan::New<v8::Array>(3);
  arr->Set(0, Nan::New(pose.vecAngularAcceleration[0]));
  arr->Set(1, Nan::New(pose.vecAngularAcceleration[1]));
  arr->Set(2, Nan::New(pose.vecAngularAcceleration[2]));
  obj->Set(Nan::New("vecAngularAcceleration").ToLocalChecked(), arr);

  obj->Set(Nan::New("result").ToLocalChecked(), Nan::New(pose.result));  

  obj->Set(Nan::New("poseIsValid").ToLocalChecked(), Nan::New(pose.poseIsValid));
  obj->Set(Nan::New("willDriftInYaw").ToLocalChecked(), Nan::New(pose.willDriftInYaw));  
  obj->Set(Nan::New("shouldApplyHeadModel").ToLocalChecked(), Nan::New(pose.shouldApplyHeadModel));  
  obj->Set(Nan::New("deviceIsConnected").ToLocalChecked(), Nan::New(pose.deviceIsConnected));

  return obj;
}
